core/src/layout.hpp

```diff
@@ -0,0 +1,185 @@
+/*
+ * Keyman is copyright (C) SIL International. MIT License.
+ *
+ * Keyman Keyboard Processor API - On-Screen Keyboard Layout Interfaces
+ */
+
+#pragma once
+
+#include <string>
+#include <map>
+#include <vector>
+
+#include "keyman_core_api.h"
+
+#if defined(__cplusplus)
+extern "C" {
+#endif
+
+/**
+ * Possible directions of a flick
+ */
+enum keyboard_layout_flick_direction {
+  /** flick up (north) */
+  n = 0,
+  /** flick down (south) */
+  s = 1,
+  /** flick right (east) */
+  e = 2,
+  /** flick left (west) */
+  w = 3,
+  /** flick up-right (north-east) */
+  ne = 4,
+  /** flick up-left (north-west) */
+  nw = 5,
+  /** flick down-right (south-east) */
+  se = 6,
+  /** flick down-left (south-west) */
+  sw = 7
+};
+
+/**
+ * key type like regular key, framekeys, deadkeys, blank, etc.
+ */
+enum keyboard_layout_key_type {
+  /** regular key */
+  normal = 0,
+  /** A 'frame' key, such as Shift or Enter */
+  special = 1,
+  /** A 'frame' key, such as Shift or Enter, which is is active, such as
+   * the shift key on a shift layer */
+  specialActive = 2,
+  /** **KeymanWeb runtime private use:** a variant of `special` with the
+   *  keyboard font rather than 'KeymanwebOsk' font */
+  customSpecial = 3,
+  /** **KeymanWeb runtime private use:** a  variant of `specialActive` with the
+   *  keyboard font rather than 'KeymanwebOsk' font. */
+  customSpecialActive = 4,
+  /** A deadkey */
+  deadkey = 8,
+  /** A key which is rendered as a blank keycap, should block any interaction */
+  blank = 9,
+  /** Renders the key only as a gap or spacer, should block any interaction */
+  spacer = 10
+};
+
+/**
+ * A key on a touch layout/on-screen keyboard
+ */
+struct keyboard_layout_key {
+  /** key id */
+  std::u16string id;  // ??? perhaps necessary for special keys, Enter, etc? or can we get that from virtualKey?
```

Comment:
```suggestion
  std::u16string id;  // TODO-WEB-CORE: perhaps necessary for special keys, Enter, etc? or can we get that from virtualKey?
```

Should change all `???` to `TODO-WEB-CORE:` so we can track these for resolution in the epic cycle.

Reply:
Done.